bagua/torch_api/contrib/utils/redis_store.py

```diff
@@ -0,0 +1,229 @@
+import socket
+import subprocess
+import time
+from bagua.torch_api.env import get_rank, get_local_rank, get_world_size, get_local_size
+
+try:
+    from redis import Redis
+except ImportError:
+    print(
+        "DEBUG: did not find redis-py. To install it, run `pip install redis` or follow instructions on its website(https://github.com/andymccurdy/redis-py)."
+    )
+    raise
+
+from typing import List, Dict, Optional, Union
+from .store import Store, ClusterStore
+import torch.distributed.distributed_c10d as c10d
+import json
+import logging
+import atexit
+
+
+try:
+    p = subprocess.Popen(
+        ["redis-server", "--version"], stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+except Exception:
+    print(
+        "DEBUG: did not find redis-server. Follow instructions on its website(https://redis.io/download) to have it installed."
+    )
+    raise
+
+
+__all__ = ["RedisStore"]
+
+_host_ip = None
+
+_global_redis_servers = []
+
+
+class RedisStore(ClusterStore):
+    """
+    A Redis-based distributed Key-Value store implementation, with ``set(...)`` and ``get(...)``` API exposed.
+
+    Args:
+        hosts (List[Dict[str, str]]): A list of redis servers, defined by a list of dict containing server host and
+            port information. New Redis instances will be spawned if ``hosts=None``.
+        cluster_mode (bool): If ``True``, data is sharded across all Redis instances. Otherwise, data is routed to a specific server.
```

Comment:
What does `routed to a specific server` mean?

Reply:
Bagua processes within the same node will store all the data in one redis instance.

If there are multiple redis instances, we enable loadbalance through a round-robin way.